ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,785 @@
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    data->xpmem_saddr = (void **) malloc(sizeof(void *) * size);
+    if (NULL == data->xpmem_saddr) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    data->xpmem_raddr = (void **) malloc(sizeof(void *) * size);
+    if (NULL == data->xpmem_raddr) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    data->rcache = (mca_rcache_base_module_t **) malloc(sizeof(mca_rcache_base_module_t *) * size);
+    if (NULL == data->rcache) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    seg_id = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE, (void *) 0666);
+    if (seg_id == -1) {
+        line = __LINE__;
+        ret = -1;
+        goto error_hndl;
+    }
+
+    ret = comm->c_coll->coll_allgather(&seg_id, sizeof(xpmem_segid_t), MPI_BYTE, data->allseg_id,
+                                       sizeof(xpmem_segid_t), MPI_BYTE, comm,
+                                       comm->c_coll->coll_allgather_module);
+
+    /* Assuming the length of rcache name is less than 50 characters */
+    char rc_name[50];
+    for (int i = 0; i < size; i++) {
+        if (rank != i) {
+            data->all_apid[i] = xpmem_get(data->allseg_id[i], XPMEM_RDWR, XPMEM_PERMIT_MODE,
+                                          (void *) 0666);
+            if (data->all_apid[i] == -1) {
+                line = __LINE__;
+                ret = -1;
+                goto error_hndl;
+            }
+            if (data->all_apid[i] == -1) {
+                line = __LINE__;
+                ret = -1;
+                goto error_hndl;
+            }
+            sprintf(rc_name, "acoll_%d_%d_%d", cid, rank, i);
+            mca_rcache_base_resources_t rcache_element
+                = {.cache_name = rc_name,
+                   .reg_data = &data->all_apid[i],
+                   .sizeof_reg = sizeof(struct acoll_xpmem_rcache_reg_t),
+                   .register_mem = mca_coll_acoll_xpmem_register,
+                   .deregister_mem = mca_coll_acoll_xpmem_deregister};
+
+            data->rcache[i] = mca_rcache_base_module_create("grdma", NULL, &rcache_element);
+            if (data->rcache[i] == NULL) {
+                ret = -1;
+                line = __LINE__;
+                goto error_hndl;
+            }
+        }
+    }
+#endif
+
+    /* temporary variables */
+    int tmp1, tmp2, tmp3 = 0;
+
+    /* shared memory structure */
+    /* first 16 * 1024 bytes (16KB) are used for the leader */
+    /* next 2* 64 * comm_size bytes are used for  sync variables */
+    /* next 8 * 1024 * comm_size are used for per_rank data (8KB per rank) */
+
+    /* offsets for the shared memory region */
+    const int leader_shm_size = 16 * 1024; 
+    const int cache_line_size = 64;
+    const int per_rank_shm_size = 8 * 1024;
```

Comment:
Thanks! Are those numbers used anywhere else? I tried to find similar values. I'm a bit concerned that changing the values here would break code elsewhere. Maybe they should be `#define`d in this header and used wherever the sizes of the buffers are relevant?

Reply:
Good point, done.

Reply:
The numbers are primarily used within *utils.h, but as you mentioned, it could be used elsewhere as well.